Feature/ifs 2483 ueberarbeitung sst authentifizierungsmanager

kollmana opened the pull request:
new pull request after including review remarks and renaming of branch due to german umlauts

kollmana committed:
docs:  typos corrected in Konzept 'Security'

kollmana committed:
docs:  typos corrected in Konzept 'Security'

kollmana committed:
docs:  AKO Test

kollmana committed:
docs:   reverted AKO Test

kollmana committed:
docs: IFS-2483: rework interface of AuthentifizierungsManager

kollmana committed:
docs: IFS-2483: rework interface of AuthentifizierungsManager

kollmana committed:
docs: IFS-2483: rework interface of AuthentifizierungsManager

kollmana committed:
docs: ISY-147 - IF3 isy-security:

Nacharbeiten nach Absprachen mit msg

kollmana committed:
docs: ISY-147 - IF3 isy-security:

Nacharbeiten nach Absprachen mit msg

kollmana committed:
docs: ISY-147 - IF3 isy-security:

Nacharbeiten nach Absprachen mit msg: Berechtigungsmanager

kollmana committed:
docs: ISY-147 - IF3 isy-security:

Nacharbeiten nach Absprachen mit msg: Berechtigungsmanager

kollmana committed:
Update isyfact-standards-doc/src/docs/antora/modules/isy-security/pages/konzept/inhalt.adoc

Co-authored-by: vonLogaJ <vonLogaJ@users.noreply.github.com>

kollmana committed:
docs: ISY-147 - IF3 isy-security:

Included review remarks

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/isy-security/images/konzept/security-schnittstellen.dn.svg:
Bitte die Methoden des Interfaces alle untereinander schreiben.

kollmana replied:
done

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/isy-security/pages/konzept/inhalt.adoc:
```diff
@@ -97,41 +96,44 @@ image::isy-security:konzept/software-architektur-berechtigungspruefung.png[align
-Der `NutzerAuthentifizierunginterceptor` ist für die Authentifizierung von technischen Nutzern zuständig und wird über eine Annotation (`@NutzerAuthentifikation`) bereitgestellt.
-Parallel zur Annotation `@NutzerAuthentifikation` existiert die Annotation `@ClientAuthentification` zur Authentifizierung von OAuth2.0 Clients.
-Dies ermöglicht die Umstellung von technische Nutzer auf OAuth2.0 Clients, ohne große Auswirkungen im Anwendungssystem zu erzeugen.
+- Der `NutzerAuthentifizierunginterceptor` ist für die Authentifizierung von technischen Nutzern zuständig und wird über eine Annotation `@NutzerAuthentifikation` bereitgestellt.
```
Ist diese im Baustein isy-security bereits implementiert bzw. wird diese bis zum Releas noch implementiert?

kollmana replied:
Siehe ISY-83

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/isy-security/pages/konzept/inhalt.adoc:
```diff
@@ -97,41 +96,44 @@ image::isy-security:konzept/software-architektur-berechtigungspruefung.png[align
-Parallel zur Annotation `@NutzerAuthentifikation` existiert die Annotation `@ClientAuthentification` zur Authentifizierung von OAuth2.0 Clients.
-Dies ermöglicht die Umstellung von technische Nutzer auf OAuth2.0 Clients, ohne große Auswirkungen im Anwendungssystem zu erzeugen.
+- Der `NutzerAuthentifizierunginterceptor` ist für die Authentifizierung von technischen Nutzern zuständig und wird über eine Annotation `@NutzerAuthentifikation` bereitgestellt.
+Parallel dazu existiert die Annotation `@ClientAuthentification` zur Authentifizierung von OAuth2.0 Clients.
```
Ist diese im Baustein isy-security bereits implementiert bzw. wird diese bis zum Releas noch implementiert?

dauna-msg replied:
Gesehen, dass sie mit isy-83 kommen werden.

kollmana replied:
Diese beiden Annotationen dienen zur Absicherung der damit annotierten Servicemethode über eine Authentifizierung des/der als Parameter aufgeführten Nutzer(s): Die fest in der gerufenen (!) Anwendung hinterlegten Credentials werden zur Authentifizierung verwendet. Evtl. übermittelte Credentials oder Access-Token der rufenden Anwendung werden ignoriert.
Die Implementierung mit isy-83 könnte bedeuten, dass es hierfür nur eine Annotation gibt, die die oauth2ClientRegistrationId als Konfigurationsschlüssel und die Methode `authentifiziere(oauth2ClientRegistrationId)` im Authentifizierungsmanager verwendet. Das Security-Konzept muss dann angepasst werden.

dauna-msg replied:
Bitte noch einmal prüfen, wie der aktuelle Stand der Annotationen ist. Ziel ist, dass beide Annotationen im Baustein implementiert sind.

kollmana replied:
Konzept angepasst an Implementierung gemäß [ISY-83 Feature/isy security auth annotation](https://github.com/IsyFact/isyfact-standards/pull/65/files#top): 
Es gibt nur eine IsyFact-Annotation `@Authenticate` mit oauth2ClientRegistrationId als Parameter und verwendet die Methode `authentifiziere(oauth2ClientRegistrationId)` im Authentifizierungsmanager.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/isy-security/pages/konzept/inhalt.adoc:
```diff
@@ -143,37 +145,65 @@ image::konzept/security-schnittstellen.dn.svg[align="center"]
+Bei erfolgreicher Authentifizierung legt die Anwendung das erhaltene Access-Token in ihrem SecurityContext ab und schickt es bei jedem Aufruf von Nachbaranwendungen mit. +
 Eine nicht erfolgreiche Authentifizierung (fachliche Ablehnung oder technisches Problem) löst eine `AuthentifizierungFehlgeschlagenException` bzw. `AuthentifizierungTechnicalException` aus.
 
+==== Authentifizierung einer externen Anwendung (über ServiceGateway):
```
Hier ist etwas durcheinander gekommen, die Authentifizierungswege werden erst später beschrieben und nicht an dieser Stelle. Ggfs. wird dieser Abschnitt nicht benötigt.

dauna-msg replied:
An dieser Stelle sollten die Methoden kurz erläutert werden, aber nicht nach interner oder SGW-Authentifizierung unterschieden.

kollmana replied:
Beschreibung (für externe Auth) gekürzt.
Aber zum einen möchte ich erläutern, wofür diese drei Auth-Methoden gedacht sind und zum anderen die Parameter erklären (speziell authentifiziereSystem mit sowohl Cl-Cred. als auch User-PW).

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/isy-security/pages/konzept/inhalt.adoc:
```diff
@@ -143,37 +145,65 @@ image::konzept/security-schnittstellen.dn.svg[align="center"]
+Der `Authentifizierungsmanager` übernimmt die programmatische Authentifizierung und Autorisierung interner und externer Anwendungen durch das IAM für den Zugang zu IsyFact-Nachbaranwendungen.
+Dafür sind folgende Szenarien zu unterscheiden, für die es jeweils eine eigene Servicemethode gibt:
+
+==== Authentifizierung einer internen Anwendung (Batch, Task, …) als eigener Schritt (direkt/intern):
```
Hier ist etwas durcheinander gekommen, die Authentifizierungswege werden erst später beschrieben und nicht an dieser Stelle.  Ggfs. wird dieser Abschnitt nicht benötigt.

An dieser Stelle sollten die Methoden kurz erläutert werden, aber nicht nach interner oder SGW-Authentifizierung unterschieden.

kollmana replied:
s.o.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/isy-security/pages/konzept/inhalt.adoc:
```diff
@@ -143,37 +145,65 @@ image::konzept/security-schnittstellen.dn.svg[align="center"]
+Die Methode `authentifiziere(oauth2ClientRegistrationId)` im Authentifizierungsmanager liest diese Authentifizierungsdaten aus der betrieblichen Konfiguration und führt damit eine OAuth2.0-konforme Authentifizierung und Autorisierung der Anwendung gegenüber dem IAM-System durch.
+Dabei wird anhand des Typs der Authentifizierungsdaten der passende Authentication-Flow mit den konfigurierten Authentifizierungsdaten verwendet. +
+Bei erfolgreicher Authentifizierung legt die Anwendung das erhaltene Access-Token in ihrem SecurityContext ab und schickt es bei jedem Aufruf von Nachbaranwendungen mit. +
 Eine nicht erfolgreiche Authentifizierung (fachliche Ablehnung oder technisches Problem) löst eine `AuthentifizierungFehlgeschlagenException` bzw. `AuthentifizierungTechnicalException` aus.
```
Werden diese Exceptions in isy-security geworfen oder ist die dortige Bezeichnung abweichend?

kollmana replied:
Das war so in Sicherheit und soll genauso sein in Security. (Da ich nicht die aktuelle Implementierung habe, kann ich nicht prüfen, ob es so ist.)

kollmana replied:
Ist dem Rotstift zum Opfer gefallen.

kollmana committed:
docs: IFS-2483: rework interface of AuthentifizierungsManager

Included PR review remarks from dauna-msg

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/isy-security/images/konzept/security-schnittstellen.dn.svg:
Bitte die Abbildung in der breite vergrößern, sodass die Methoden jeweils in einer Zeile stehen und nicht umgebrochen werden müssen.

kollmana replied:
Der Text darin wird dann ziemlich klein, ist aber noch lesbar.

kollmana committed:
docs: IFS-2483: rework interface of AuthentifizierungsManager

Included PR review remarks from dauna-msg

kollmana committed:
docs: IFS-2483: rework interface of AuthentifizierungsManager

Seperate methods for client & system authentication without BHKNZ (as in ISY-147)

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/isy-security/pages/konzept/inhalt.adoc:
```diff
@@ -121,17 +121,18 @@ Der `RolePrivilegeGrantedAuthoritiesConverter` ist für die Konvertierung der vo
+Im Interceptor `AuthenticateInterceptor` erfolgt eine Authentifizierung über die Methode `authentifiziere(oauth2ClientRegistrationId)` im Authentifizierungsmanager.
+Der Service wird dann im Namen und mit den Rechten des Ressourcenbesitzers ausgeführt.
+
+Hinweis: Die Annotation `@Authenticate` löst die Annotation `@NutzerAuthentifikation` ab und erweitert sie um die Funktion, auch hinterlegte OAuth-2-Clients authentifizieren zu können.
```
Werden nicht-Oauth2.0-Clients, also die "alten" techn. Nutzer, auch darüber authentifiziert?

kollmana replied:
Da die Annotation als Parameter ‚oauth2ClientRegistrationId‘ hat und intern die Methode authentifiziere(oauth2ClientRegistrationId)verwendet wird, werden die selben Arten von Benutzern auf die selbe Weise authentifiziert: über die oauth2ClientRegistrationId werden die in der Konfig hinterlegten Authentifizierungsdaten ausgelesen und je nach Typ (techn.Nutzer = Resource Owner bzw. Oauth2.0-Client = Client Cred.) der entsprechende Authentifizierungsflow gewählt. 

Das dachte ich in der Beschreibung des AuthenticateInterceptors beschrieben zu haben und auf die Annotation bezogen zu haben. 
Mit ‚erweitert‘ und ‚auch‘ wollte ich ausdrücken, dass @Authenticate zusätzlich zu der bisher mit @NutzerAuthentifikation möglichen Authentifizierung von nicht-Oauth2.0-Clients, also die "alten" techn. Nutzer auch Oauth2.0-Clients authentifiziert werden können – bzw. Benutzer, die in der Konfig unter der oauth2ClientRegistrationId abgelegt sind, mit dem zu verwendenden Auth-Typ.

kollmana commented:
Aktuelle Textstelle:

Eine Absicherung von Services auf Ebene des Ressourcenbesitzers (Resource Owner), also eines technischen Nutzers bzw. OAuth-2-Client bietet die IsyFact-Annotation @Authenticate. Dieser Annotation muss noch die oauth2ClientRegistrationId mitgegeben werden, unter der der Ressourcenbesitzer in der betrieblichen Konfiguration registriert ist. Im Interceptor AuthenticateInterceptor erfolgt eine Authentifizierung über die Methode authentifiziere(oauth2ClientRegistrationId) im Authentifizierungsmanager. Der Service wird dann im Namen und mit den Rechten des Ressourcenbesitzers ausgeführt.

Hinweis: Die Annotation @Authenticate löst die Annotation @NutzerAuthentifikation ab und erweitert sie um die Funktion, auch hinterlegte OAuth-2-Clients authentifizieren zu können.

Du fragtest: 'Werden nicht-Oauth2.0-Clients, also die "alten" techn. Nutzer, auch darüber authentifiziert?'

Da die Annotation als Parameter 'oauth2ClientRegistrationId' hat und intern die Methode authentifiziere(oauth2ClientRegistrationId)verwendet wird, werden die selben Arten von Benutzern auf die selbe Weise authentifiziert: über die oauth2ClientRegistrationId werden die in der Konfig hinterlegten Authentifizierungsdaten ausgelesen und je nach Typ (techn.Nutzer = Resource Owner bzw. Oauth2.0-Client = Client Cred.) der entsprechende Authentifizierungsflow gewählt.

Das dachte ich in der Beschreibung des AuthenticateInterceptors beschrieben zu haben und auf die Annotation bezogen zu haben.
Mit 'erweitert' und 'auch' wollte ich ausdrücken, dass @Authenticate zusätzlich zu der bisher mit @NutzerAuthentifikation möglichen Authentifizierung von nicht-Oauth2.0-Clients, also die "alten" techn. Nutzer auch Oauth2.0-Clients authentifiziert werden können - bzw. Benutzer, die in der Konfig unter der oauth2ClientRegistrationId abgelegt sind, mit dem zu verwendenden Auth-Typ.

Wenn mein Text falsch oder nicht verständlich genug ist, bitte ich Dich um Verbesserungsvorschläge.

Gruß
Andreas Kollmann
